interfaces/builtin/compatibility.go

```diff
@@ -0,0 +1,180 @@
+	// can have Min!=Max.
+	Values []CompatRange
+}
+
+type CompatField struct {
+	Dimensions []CompatDimension
+}
+
+const (
+	maxDimensions = 3
+	maxIntegers   = 3
+)
+
+var (
+	tagRegexp = regexp.MustCompile(`^[a-z][a-z0-9]{0,31}$`)
+	// 0 or a number that starts with 1-9 with a maximum of 8 digits
+	intRegexp      = regexp.MustCompile(`^(0|[1-9][0-9]{0,7})$`)
+	intRangeRegexp = regexp.MustCompile(`^\((0|[1-9][0-9]{0,7})\.\.(0|[1-9][0-9]{0,7})\)$`)
+)
+
+func isRealCompatRange(cr CompatRange) bool {
+	return cr.Min < cr.Max
+}
+
+func appendDimension(dimensions []CompatDimension, dimension *CompatDimension) []CompatDimension {
+	if dimension != nil {
+		if len(dimension.Values) == 0 {
+			// If we have just a tag, that is equivalent to <tag>-0
+			dimension.Values = []CompatRange{{0, 0}}
+		}
+		dimensions = append(dimensions, *dimension)
+	}
+	return dimensions
+}
+
+func appendRange(ranges []CompatRange, rg CompatRange) ([]CompatRange, error) {
+	numVal := len(ranges)
+	if numVal > 0 && isRealCompatRange(ranges[numVal-1]) {
+		return nil, fmt.Errorf("ranges only allowed at the end of compatibility field")
+	}
+	if numVal == maxIntegers {
+		return nil, fmt.Errorf("only %d integer/integer ranges allowed per dimension in compatibility field", maxIntegers)
+	}
+	return append(ranges, rg), nil
+}
+
+func parseIntegerRange(token, compat string) (rg *CompatRange, err error) {
+	if intRegexp.MatchString(token) {
+		// 8 decimals fit 32 bits
+		min, err := strconv.ParseUint(token, 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		// Integers are stored as ranges where min==max
+		rg = &CompatRange{uint(min), uint(min)}
+	} else {
+		matches := intRangeRegexp.FindStringSubmatch(token)
+		// Not a tag, integer or integer range
+		if len(matches) == 0 {
+			return nil, fmt.Errorf("invalid tag %q in compatibility field %q",
+				token, compat)
+		}
+		// 8 decimals fit 32 bits
+		min, err := strconv.ParseUint(matches[1], 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		max, err := strconv.ParseUint(matches[2], 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		if min > max {
+			return nil, fmt.Errorf("invalid range %q in compatibility field %q",
+				token, compat)
+		}
+		rg = &CompatRange{uint(min), uint(max)}
+	}
+	return rg, nil
+}
+
+// decodeCompatField decodes a compatibility string, which is a sequence of
+// dimensions separated by dashes. Each dimension consist of an alphanumerical
+// descriptor followed by a dash-separated series of integer or integer ranges.
+func decodeCompatField(compat string) (*CompatField, error) {
+	tokens := strings.Split(compat, "-")
+	dimensions := []CompatDimension{}
+	var currentDimension *CompatDimension
+	for _, t := range tokens {
+		if tagRegexp.MatchString(t) {
+			if len(dimensions)+1 >= maxDimensions {
+				return nil, fmt.Errorf("only %d dimensions allowed in compatibility field: %q",
+					maxDimensions, compat)
+			}
+			dimensions = appendDimension(dimensions, currentDimension)
```

Comment:
are we missing the check that the same tag is not reused multiple times?

Reply:
I'm doubtful that it would make sense to repeat,   as would a-2-a-3 mean  vs a-3-a-2 ? it's extremely confusing